File: android/app/src/test/java/com/hawksnest/core/logic/EntityVisibilityTest.kt
package com.hawksnest.core.logic

import com.hawksnest.core.ha.HassEntity
import kotlinx.serialization.json.buildJsonObject
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

/** Ports `entityVisibility.test.ts` 1:1. */
class EntityVisibilityTest {

    private val categories = mapOf(
        "sensor.back_battery" to "diagnostic",
        "number.back_volume" to "config",
    )

    private fun ent(id: String): HassEntity =
        HassEntity(entityId = id, state = "on", attributes = buildJsonObject {})

    @Test
    fun `keeps real controls and signals`() {
        for (id in listOf(
            "lock.front_door",
            "light.basement",
            "camera.back",
            "binary_sensor.back_motion",
            "switch.back_siren",
            "alarm_control_panel.mfa_alarm",
        )) {
            assertTrue(isPrimaryEntity(id, categories), id)
        }
    }

    @Test
    fun `drops HA config and diagnostic entities`() {
        assertFalse(isPrimaryEntity("sensor.back_battery", categories))
        assertFalse(isPrimaryEntity("number.back_volume", categories))
    }

    @Test
    fun `drops untagged ring-mqtt housekeeping entities by suffix`() {
        for (id in listOf(
            "sensor.back_last_activity",
            "sensor.back_info",
            "camera.back_event_stream",
            "camera.back_live_stream",
            "select.back_event_select",
            "select.back_bypass_mode",
            "select.back_chirp_tone",
            "camera.back_snapshot",
            "camera.back_live_view",
        )) {
            assertTrue(isNoiseEntity(id), id)
            assertFalse(isPrimaryEntity(id, emptyMap()), id)
        }
    }

    @Test
    fun `does not treat a real control as noise`() {
        assertFalse(isNoiseEntity("lock.front_door"))
        assertFalse(isNoiseEntity("binary_sensor.back_motion"))
    }

    @Test
    fun `primaryEntities filters a list`() {
        val list = listOf(
            ent("lock.front_door"),
            ent("sensor.back_last_activity"),
            ent("sensor.back_battery"),
            ent("light.basement"),
        )
        assertEquals(
            listOf("lock.front_door", "light.basement"),
            primaryEntities(list, categories).map { it.entityId },
        )
    }
}

File: src/lib/__tests__/entityVisibility.test.ts
import { describe, it, expect } from "vitest";
import {
  isPrimaryEntity,
  isNoiseEntity,
  primaryEntities,
} from "../entityVisibility";
import type { HassEntity } from "../ha";

const ent = (id: string): HassEntity => ({ entity_id: id, state: "on", attributes: {} });

describe("isPrimaryEntity", () => {
  const categories = {
    "sensor.back_battery": "diagnostic",
    "number.back_volume": "config",
  };

  it("keeps real controls and signals", () => {
    for (const id of [
      "lock.front_door",
      "light.basement",
      "camera.back",
      "binary_sensor.back_motion",
      "switch.back_siren",
      "alarm_control_panel.mfa_alarm",
    ]) {
      expect(isPrimaryEntity(id, categories)).toBe(true);
    }
  });

  it("drops HA config/diagnostic entities", () => {
    expect(isPrimaryEntity("sensor.back_battery", categories)).toBe(false);
    expect(isPrimaryEntity("number.back_volume", categories)).toBe(false);
  });

  it("drops untagged ring-mqtt housekeeping entities by suffix", () => {
    for (const id of [
      "sensor.back_last_activity",
      "sensor.back_info",
      "camera.back_event_stream",
      "camera.back_live_stream",
      "select.back_event_select",
      "select.back_bypass_mode",
      "select.back_chirp_tone",
      "camera.back_snapshot",
      "camera.back_live_view",
    ]) {
      expect(isNoiseEntity(id)).toBe(true);
      expect(isPrimaryEntity(id, {})).toBe(false);
    }
  });

  it("does not treat a real control as noise", () => {
    expect(isNoiseEntity("lock.front_door")).toBe(false);
    expect(isNoiseEntity("binary_sensor.back_motion")).toBe(false);
  });

  it("primaryEntities filters a list", () => {
    const list = [
      ent("lock.front_door"),
      ent("sensor.back_last_activity"),
      ent("sensor.back_battery"),
      ent("light.basement"),
    ];
    expect(primaryEntities(list, categories).map((e) => e.entity_id)).toEqual([
      "lock.front_door",
      "light.basement",
    ]);
  });
});
